examples/fnirs/plot_classif_fnirs.py

```diff
@@ -0,0 +1,482 @@
+                    )
+
+            # Add the flattening transformer at the end of the pipeline
+            block_pipeline.steps.append(('flatten', FlattenTransformer()))
+
+            transformers.append((block_name, block_pipeline, [block_name]))
+
+        # create the columncransformer with per-block pipelines
+        self.preprocessor = ColumnTransformer(transformers)
+
+        # Prepare the DataFrame
+        X_df = self._prepare_dataframe(X)
+
+        # Fit the preprocessor
+        self.preprocessor.fit(X_df)
+
+        return self
+
+    def transform(self, X):
+        """Estimate block kernel or covariance matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_times)
+            Multi-channel time series.
+
+        Returns
+        -------
+        M : ndarray, shape (n_matrices, n_channels, n_channels)
+            Block diagonal matrices (kernel or covariance matrices).
+        """
+        check_is_fitted(self, 'preprocessor')
+
+        # make the df wheren each block is 1 column
+        X_df = self._prepare_dataframe(X)
+
+        # Transform the data
+        transformed_blocks = []
+        data_transformed = self.preprocessor.transform(X_df)
+
+        # calculate the number of features per block
+        features_per_block = [size * size for size in self.blocks]
+
+        # compute the indices where to split the data
+        split_indices = np.cumsum(features_per_block)[:-1]
+
+        # split the data into flattened blocks
+        blocks_flat = np.split(data_transformed, split_indices, axis=1)
+
+        # reshape each block back to its original shape
+        for i, block_flat in enumerate(blocks_flat):
+            size = self.blocks[i]
+            block = block_flat.reshape(-1, size, size)
+            transformed_blocks.append(block)
+
+        # Reconstruct the block diagonal matrices
+        n_samples = X.shape[0]
+        M_matrices = []
+        for i in range(n_samples):
+            blocks = [Xt[i] for Xt in transformed_blocks]
+            M_full = self._block_diag(blocks)
+            M_matrices.append(M_full)
+        return np.array(M_matrices)
+
+    def _prepare_dataframe(self, X):
+        """Converts the data into a df with eac hblock as column."""
+        data_dict = {}
+        for i, (start, end) in enumerate(self.block_indices):
+            data_dict[self.block_names[i]] = list(X[:, start:end, :])
+        return pd.DataFrame(data_dict)
+
+    @staticmethod
+    def _block_diag(matrices):
+        """Construct a block diagonal matrix from a list of square matrices."""
+        shapes = [m.shape[0] for m in matrices]
+        total_size = sum(shapes)
+        result = np.zeros((total_size, total_size), dtype=matrices[0].dtype)
+        start = 0
+        for m in matrices:
+            end = start + m.shape[0]
+            result[start:end, start:end] = m
+            start = end
+        return result
```

Comment:
I may have said it before but did you consider using https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.block_diag.html ?

Reply:
thanks for the pointer. This simplifies the block diagonal computation a lot